drivers/pca9685/pca9685.c

```diff
@@ -0,0 +1,382 @@
+            case PWM_LEFT: on = 0;
+                           off = on + duty;
+                           break;
+            case PWM_CENTER: on = (PCA9685_RESOLUTION - duty) >> 1;
+                             off = on + duty;
+                             break;
+            case PWM_RIGHT: off = PCA9685_RESOLUTION - 1;
+                            on = off - duty;
+                            break;
+            default: return;
+        }
+    }
+
+    DEBUG_DEV("on=%u off=%u", dev, on, off);
+
+    if (chn == PCA9685_CHANNEL_NUM) {
+        EXEC(_write_word(dev, PCA9685_REG_ALL_LED_ON, on));
+        EXEC(_write_word(dev, PCA9685_REG_ALL_LED_OFF, off));
+    }
+    else {
+        EXEC(_write_word(dev, PCA9685_REG_LED_ON(chn), on));
+        EXEC(_write_word(dev, PCA9685_REG_LED_OFF(chn), off));
+    }
+}
+
+void pca9685_pwm_poweron(pca9685_t *dev)
+{
+    ASSERT_PARAM(dev != NULL);
+    DEBUG_DEV("", dev);
+
+    uint8_t byte;
+    /* read MODE1 register */
+    EXEC(_read(dev, PCA9685_REG_MODE1, &byte, 1));
+
+    /* check if RESTART bit is 1 */
+    if (byte & PCA9685_MODE1_RESTART) {
+        /* clear the SLEEP bit */
+        byte &= ~PCA9685_MODE1_SLEEP;
+        EXEC(_write(dev, PCA9685_REG_MODE1, &byte, 1));
+        /* allow 500 us for oscilator to stabilize */
+        xtimer_usleep(500);
+        /* clear the RESTART bit to start all PWM channels*/
+        EXEC(_update(dev, PCA9685_REG_MODE1, PCA9685_MODE1_RESTART, 1));
+    }
+    else {
+        EXEC(_update(dev, PCA9685_REG_MODE1, PCA9685_MODE1_SLEEP, 0));
+        /* allow 500 us for oscilator to stabilize */
+        xtimer_usleep(500);
+        /* clear the RESTART bit to start all PWM channels*/
+        EXEC(_update(dev, PCA9685_REG_MODE1, PCA9685_MODE1_RESTART, 1));
+    }
+
+    if (dev->params.oe_pin != GPIO_UNDEF) {
+        gpio_clear(dev->params.oe_pin);
+    }
+
+    dev->powered_on = true;
+}
+
+void pca9685_pwm_poweroff(pca9685_t *dev)
+{
+    ASSERT_PARAM(dev != NULL);
+    DEBUG_DEV("", dev);
+
+    if (dev->params.oe_pin != GPIO_UNDEF) {
+        gpio_set(dev->params.oe_pin);
+    }
+
+    /* set sleep mode */
+    EXEC(_update(dev, PCA9685_REG_MODE1, PCA9685_MODE1_SLEEP, 1));
+
+    dev->powered_on = false;
+}
+
+/** Functions for internal use only */
+
+static int _is_available(const pca9685_t *dev)
+{
+    uint8_t byte;
+
+    /* simply tests to read */
+    return _read(dev, PCA9685_REG_MODE1, &byte, 1);
+}
+
+static int _init(pca9685_t *dev)
+{
+    /* set Auto-Increment flag */
+    EXEC_RET(_update(dev, PCA9685_REG_MODE1, PCA9685_MODE1_AI, 1));
+
+    /* switch off all channels */
+    EXEC_RET(_write_word(dev, PCA9685_REG_ALL_LED_OFF, PCA9685_ALL_LED_OFF));
+
+    /* set Auto-Increment flag */
+    uint8_t byte = 0;
+    _set_reg_bit(&byte, PCA9685_MODE2_OUTDRV, dev->params.inv);
+    _set_reg_bit(&byte, PCA9685_MODE2_OUTDRV, dev->params.out_drv);
+    _set_reg_bit(&byte, PCA9685_MODE2_OUTNE, dev->params.out_ne);
```

Comment:
```suggestion
    _set_reg_bit(&byte, PCA9685_MODE2_OUTNE0, dev->params.out_ne & 0x1);
    _set_reg_bit(&byte, PCA9685_MODE2_OUTNE1, dev->params.out_ne & 0x2);
```

Maybe this is too hacky, but it world make the set function much simpler.
Another solution world be setting the already shifted bits.

(I can live with the while loop too, I'd just rather see if we can do without)

Reply:
Hm, this executed only once. The while loop is not a big thing. I would prefer to leave it as it is, for the sake of clarity. We could think about not to define only a bit mask but also a bit shift value for each for register field. Another way would be to define structures for all registers.

But, since `_set_reg_bit` is only used during `_init`, there should be no performance problem.

But, I realized a real real problem.
```diff
-  _set_reg_bit(&byte, PCA9685_MODE2_OUTDRV, dev->params.inv);
+  _set_reg_bit(&byte, PCA9685_MODE2_INVERT, dev->params.inv);
```